cobalt/browser/metrics/cobalt_metrics_service_client_test.cc

```diff
@@ -12,36 +12,377 @@
+  void CallInitialize() { Initialize(); }
+
+  StrictMock<MockMetricsService>* mock_metrics_service() const {
+    return mock_metrics_service_;
+  }
+  StrictMock<MockCobaltMetricsLogUploader>* mock_log_uploader() const {
+    return mock_log_uploader_;
+  }
+
+ private:
+  raw_ptr<StrictMock<MockMetricsService>> mock_metrics_service_ = nullptr;
+  raw_ptr<StrictMock<MockCobaltMetricsLogUploader>> mock_log_uploader_ =
+      nullptr;
+  bool on_application_not_idle_internal_called_ = false;
+};
+
+class CobaltMetricsServiceClientTest : public ::testing::Test {
+ protected:
+  CobaltMetricsServiceClientTest() { mojo::core::Init(); }
+
   void SetUp() override {
-    // This set up (and the corresponding include) are only to support the
-    // Add/RemoveActionCallback() calls inside CobaltMetricsServiceClient's
-    // MetricsService.
-    // TODO(b/372559349): Mock here said MetricsService, add expectations to its
-    // methods and remove this call and include file.
+    ASSERT_TRUE(temp_dir_.CreateUniqueTempDir());
+    path_override_ = std::make_unique<base::ScopedPathOverride>(
+        content::SHELL_DIR_USER_DATA, temp_dir_.GetPath());
+
+    // Register all metric-related prefs, otherwise tests crash when calling
+    // upstream metrics code.
+    metrics::MetricsService::RegisterPrefs(prefs_.registry());
+    prefs_.registry()->RegisterBooleanPref(
+        metrics::prefs::kMetricsReportingEnabled, false);
+
+    // Make sure user_metrics.cc has a g_task_runner.
     base::SetRecordActionTaskRunner(
         task_environment_.GetMainThreadTaskRunner());
+
+    // The client under test receives MetricsStateManager, so create one.
+    // A real CobaltEnabledStateProvider is simple enough.
+    enabled_state_provider_ =
+        std::make_unique<CobaltEnabledStateProvider>(&prefs_);
+    metrics_state_manager_ = metrics::MetricsStateManager::Create(
+        &prefs_, enabled_state_provider_.get(), std::wstring(),
+        temp_dir_.GetPath(), metrics::StartupVisibility::kForeground);
+    ASSERT_THAT(metrics_state_manager_, NotNull());
+
+    synthetic_trial_registry_ =
+        std::make_unique<variations::SyntheticTrialRegistry>();
+
+    // Instantiate the test client and call Initialize to trigger mock creation.
+    // This simulates the two-phase initialization of the static Create().
+    client_ = std::make_unique<TestCobaltMetricsServiceClient>(
+        metrics_state_manager_.get(), synthetic_trial_registry_.get(), &prefs_);
+    client_->CallInitialize();  // This will use the overridden factory methods.
+
+    ASSERT_THAT(client_->mock_metrics_service(), NotNull());
+    ASSERT_THAT(client_->mock_log_uploader(), NotNull());
   }
 
- protected:
-  base::test::SingleThreadTaskEnvironment task_environment_;
+  void TearDown() override {
+    client_.reset();
+    synthetic_trial_registry_.reset();
+    metrics_state_manager_.reset();
+    enabled_state_provider_.reset();
+    path_override_.reset();
```

Comment:
nit: Why is this specific, as opposed to letting the dtor do its
job? Does this need to happen before e.g. `task_environment_` does
a last run..? If it turns out to be needed, a comment might be
good for future reference.

Reply:
Not needed! Not sure what I was thinking. Removed in a local change that'll push shortly.